tractor/_spawn.py

```diff
@@ -356,23 +404,53 @@ async def new_proc(
             # awaited and reported upwards to the supervisor.
             await actor_nursery._join_procs.wait()
 
+        finally:
+            # XXX: in the case we were cancelled before the sub-proc
+            # registered itself back we must be sure to try and clean
+            # any process we may have started.
+
+            reaping_cancelled = False
+            cancel_scope = None
+
             if portal in actor_nursery._cancel_after_result_on_exit:
-                cancel_scope = await nursery.start(
-                    cancel_on_completion, portal, subactor, errors)
+                try:
+                    # async with trio.open_nursery() as n:
+                    # n.cancel_scope.shield = True
+                    cancel_scope = await nursery.start(
+                        cancel_on_completion,
+                        portal,
+                        subactor,
+                        errors
+                    )
+                except trio.Cancelled:
+                    # if the reaping task was cancelled we may have hit
+                    # a race where the subproc disconnected before we
+                    # could send it a message to cancel (classic 2 generals)
+                    # in that case, wait shortly then kill the process.
+                    reaping_cancelled = True
+
+                    if proc.is_alive():
+                        with trio.move_on_after(0.1) as cs:
+                            cs.shield = True
+                            await proc_waiter(proc)
+
+                        if cs.cancelled_caught:
+                            proc.terminate()
+
+            if not reaping_cancelled:
+                if proc.is_alive():
```

Comment:
So maybe we can just rely on the proc being dead here?
I guess it would make more sense to poll and then check?

If that works we can probably drop the flag set and check.